src/main/docs/man/man3/DragonGamepad.3
.TH "DragonGamepad" 3 "Thu Oct 31 2019" "2020 Template Project" \" -*- nroff -*-
.ad l
.nh
.SH NAME
DragonGamepad
.SH SYNOPSIS
.br
.PP
.PP
\fC#include <DragonGamepad\&.h>\fP
.PP
Inherits \fBIDragonGamePad\fP\&.
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "\fBDragonGamepad\fP (int port)"
.br
.ti -1c
.RI "\fB~DragonGamepad\fP ()"
.br
.ti -1c
.RI "bool \fBIsButtonPressed\fP (\fBBUTTON_IDENTIFIER\fP button) const override"
.br
.ti -1c
.RI "void \fBSetAxisDeadband\fP (\fBAXIS_IDENTIFIER\fP axis, \fBAXIS_DEADBAND\fP type)"
.br
.RI "Method: SetAxisDeadband Description: Specify what deadband behavior is desired such as none, standard, standard with scaling\&. Returns: void "
.ti -1c
.RI "void \fBSetAxisProfile\fP (\fBAXIS_IDENTIFIER\fP axis, \fBAXIS_PROFILE\fP curve)"
.br
.ti -1c
.RI "void \fBSetAxisScale\fP (\fBAXIS_IDENTIFIER\fP axis, float scaleFactor)"
.br
.ti -1c
.RI "void \fBSetButtonMode\fP (\fBBUTTON_IDENTIFIER\fP button, \fBBUTTON_MODE\fP mode)"
.br
.RI "Method: SetButtonMode Description: Specify how the button should behave\&. Examples include (but not limited to): "
.ti -1c
.RI "double \fBGetRawAxis\fP (\fBAXIS_IDENTIFIER\fP axis)"
.br
.ti -1c
.RI "bool \fBIsRawButtonPressed\fP (\fBBUTTON_IDENTIFIER\fP button)"
.br
.ti -1c
.RI "bool \fBIsRawButtonPressed\fP (\fBANALOG_BUTTON_IDENTIFIER\fP button)"
.br
.ti -1c
.RI "double \fBGetAxisValue\fP (\fBAXIS_IDENTIFIER\fP axis) const override"
.br
.ti -1c
.RI "bool \fBWasButtonPressed\fP (\fBBUTTON_IDENTIFIER\fP button) const override"
.br
.RI "Method: WasButtonPressed Description: Read whether the button was pressed since the last query\&. This is only valid for digital buttons (normal buttons and bumpers)\&. "
.ti -1c
.RI "bool \fBWasButtonReleased\fP (\fBBUTTON_IDENTIFIER\fP button) const override"
.br
.RI "Method: WasButtonReleased Description: Read whether the button was released since the last query\&. This is only valid for digital buttons (normal buttons and bumpers)\&. "
.ti -1c
.RI "int \fBGetPOVValue\fP ()"
.br
.in -1c
.SS "Additional Inherited Members"
.SH "Detailed Description"
.PP 
Definition at line 18 of file DragonGamepad\&.h\&.
.SH "Constructor & Destructor Documentation"
.PP 
.SS "DragonGamepad::DragonGamepad (int port)"

.PP
Definition at line 21 of file DragonGamepad\&.cpp\&.
.SS "DragonGamepad::~DragonGamepad ()"

.PP
Definition at line 75 of file DragonGamepad\&.cpp\&.
.SH "Member Function Documentation"
.PP 
.SS "double DragonGamepad::GetAxisValue (\fBAXIS_IDENTIFIER\fP axis) const\fC [override]\fP, \fC [virtual]\fP"

.PP
 Method: GetAxisValue Description: Return the current value (between -1\&.0 and 1\&.0) for the requested axis\&. 
.SS "Returns:     double   - current axis value"

.PP
Implements \fBIDragonGamePad\fP\&.
.PP
Definition at line 82 of file DragonGamepad\&.cpp\&.
.SS "int DragonGamepad::GetPOVValue ()"

.PP
Definition at line 241 of file DragonGamepad\&.cpp\&.
.SS "double DragonGamepad::GetRawAxis (\fBAXIS_IDENTIFIER\fP axis)"

.PP
Definition at line 199 of file DragonGamepad\&.cpp\&.
.SS "bool DragonGamepad::IsButtonPressed (\fBBUTTON_IDENTIFIER\fP button) const\fC [override]\fP, \fC [virtual]\fP"

.PP
 Method: IsButtonPressed Description: Return whether the requested button is selected (true) or not (false) Returns: bool true - button is pressed 
.SS "false - button is not pressed"

.PP
Implements \fBIDragonGamePad\fP\&.
.PP
Definition at line 145 of file DragonGamepad\&.cpp\&.
.SS "bool DragonGamepad::IsRawButtonPressed (\fBANALOG_BUTTON_IDENTIFIER\fP button)"

.PP
Definition at line 229 of file DragonGamepad\&.cpp\&.
.SS "bool DragonGamepad::IsRawButtonPressed (\fBBUTTON_IDENTIFIER\fP button)"

.PP
Definition at line 216 of file DragonGamepad\&.cpp\&.
.SS "void DragonGamepad::SetAxisDeadband (\fBAXIS_IDENTIFIER\fP axis, \fBAXIS_DEADBAND\fP type)\fC [virtual]\fP"

.PP
Method: SetAxisDeadband Description: Specify what deadband behavior is desired such as none, standard, standard with scaling\&. Returns: void ================================================================================== 
.SH "</summary>"
.PP

.PP
\fBParameters\fP
.RS 4
\fItype\fP \fI - axis to modify \fI - deadband option \fP\fP
.RE
.PP

.PP
Implements \fBIDragonGamePad\fP\&.
.PP
Definition at line 154 of file DragonGamepad\&.cpp\&.
.SS "void DragonGamepad::SetAxisProfile (\fBAXIS_IDENTIFIER\fP axis, \fBAXIS_PROFILE\fP curve)\fC [virtual]\fP"

.PP
 Method: SetProfile Description: Specify the profile curve used for setting the sensitivity of the axis\&. By default, this is linear, but a cubic curve would give more control when the axis is barely pressed (e\&.g\&. if it were moved 50% forward, instead of returning 0\&.5, it would return 0\&.5 * 0\&.5 * 0\&.5 or \&.125, but when the axis was moved all the way forward, it would return the same value -- 1\&.0\&. Since it is cubed, it retains the sign\&.
.PP
This affects values returned from GetAxis calls\&. 
.SS "Returns:     void"

.PP
Implements \fBIDragonGamePad\fP\&.
.PP
Definition at line 166 of file DragonGamepad\&.cpp\&.
.SS "void DragonGamepad::SetAxisScale (\fBAXIS_IDENTIFIER\fP axis, float scaleFactor)\fC [virtual]\fP"

.PP
 Method: SetScale Description: Scale the returned value to a range between the specified negative scale factor and the scale factor\&. This is used to reduce the maximum value returned\&.
.PP
This affects values returned from GetAxis calls\&. 
.SS "Returns:     void"

.PP
Implements \fBIDragonGamePad\fP\&.
.PP
Definition at line 178 of file DragonGamepad\&.cpp\&.
.SS "void DragonGamepad::SetButtonMode (\fBBUTTON_IDENTIFIER\fP button, \fBBUTTON_MODE\fP mode)\fC [virtual]\fP"

.PP
Method: SetButtonMode Description: Specify how the button should behave\&. Examples include (but not limited to): 
.IP "\(bu" 2
pressed / not pressed
.IP "\(bu" 2
toggle Returns: void
.PP

.PP
\fBParameters\fP
.RS 4
\fImode\fP \fI - button to check \fI - button behavior \fP\fP
.RE
.PP

.PP
Implements \fBIDragonGamePad\fP\&.
.PP
Definition at line 190 of file DragonGamepad\&.cpp\&.
.SS "bool DragonGamepad::WasButtonPressed (\fBBUTTON_IDENTIFIER\fP button) const\fC [override]\fP, \fC [virtual]\fP"

.PP
Method: WasButtonPressed Description: Read whether the button was pressed since the last query\&. This is only valid for digital buttons (normal buttons and bumpers)\&. 
.PP
Implements \fBIDragonGamePad\fP\&.
.PP
Definition at line 248 of file DragonGamepad\&.cpp\&.
.SS "bool DragonGamepad::WasButtonReleased (\fBBUTTON_IDENTIFIER\fP button) const\fC [override]\fP, \fC [virtual]\fP"

.PP
Method: WasButtonReleased Description: Read whether the button was released since the last query\&. This is only valid for digital buttons (normal buttons and bumpers)\&. 
.PP
Implements \fBIDragonGamePad\fP\&.
.PP
Definition at line 256 of file DragonGamepad\&.cpp\&.

.SH "Author"
.PP 
Generated automatically by Doxygen for 2020 Template Project from the source code\&.


src/main/cpp/controllers/DragonGamepad.cpp
/*----------------------------------------------------------------------------*/
/* Copyright (c) 2018 FIRST. All Rights Reserved.                             */
/* Open Source Software - may be modified and shared by FRC teams. The code   */
/* must be accompanied by the FIRST BSD license file in the root directory of */
/* the project.                                                               */
/*----------------------------------------------------------------------------*/

#include <controllers/DragonGamepad.h>

#include <controllers/axis/IDeadband.h>
#include <controllers/axis/IProfile.h>

#include <controllers/button/DigitalButton.h>
#include <controllers/axis/AnalogAxis.h>
#include <controllers/button/AnalogButton.h>

#include <frc/GenericHID.h>
#include <frc/Joystick.h>

DragonGamepad::DragonGamepad
(
    int port
)  : m_gamepad( new frc::Joystick(port)),
m_axis(),
m_button(),
m_count(),
m_axisScale(),
m_axisInversionFactor(),
m_axisProfile()
{
    for ( auto inx=0; inx<m_maxAxis; ++inx )
		{
			m_axisScale[inx] = 1.0;
			m_axisInversionFactor[inx] = 1.0;
			m_axisProfile[inx] = LINEAR;
		}
    //Create Axis objects
    m_axis[GAMEPAD_AXIS_16] = new AnalogAxis(m_gamepad, 0,false);
    m_axis[GAMEPAD_AXIS_17] = new AnalogAxis(m_gamepad, 0,false);
    m_axis[GAMEPAD_DIAL_22] = new AnalogAxis(m_gamepad, GAMEPAD_DIAL_22,false);
    m_axis[GAMEPAD_DIAL_23] = new AnalogAxis(m_gamepad, GAMEPAD_DIAL_23,false);
    m_axis[GAMEPAD_DIAL_24] = new AnalogAxis(m_gamepad, GAMEPAD_DIAL_24,false);
    m_axis[GAMEPAD_DIAL_25] = new AnalogAxis(m_gamepad, GAMEPAD_DIAL_25,false);
    m_axis[GAMEPAD_DIAL_26] = new AnalogAxis(m_gamepad, GAMEPAD_DIAL_26,false);
    m_axis[GAMEPAD_DIAL_27] = new AnalogAxis(m_gamepad, GAMEPAD_DIAL_27,false);
    m_axis[LEFT_ANALOG_BUTTON_AXIS] = new AnalogAxis(m_gamepad, LEFT_BUTTON_AXIS_ID,false );
    m_axis[RIGHT_ANALOG_BUTTON_AXIS] = new AnalogAxis(m_gamepad, RIGHT_BUTTON_AXIS_ID,false);

    //Create Button objects
    m_analogButtons[GAMEPAD_BUTTON_1] = new AnalogButton(m_axis[LEFT_ANALOG_BUTTON_AXIS], BUTTON_1_LOWERBAND,BUTTON_1_UPPERBAND);
    m_analogButtons[GAMEPAD_BUTTON_2] = new AnalogButton( m_axis[RIGHT_ANALOG_BUTTON_AXIS], BUTTON_2_LOWERBAND,BUTTON_2_UPPERBAND);
    m_analogButtons[GAMEPAD_BUTTON_3] = new AnalogButton(m_axis[LEFT_ANALOG_BUTTON_AXIS], BUTTON_3_LOWERBAND,BUTTON_3_UPPERBAND);
    m_analogButtons[GAMEPAD_BUTTON_4] = new AnalogButton( m_axis[RIGHT_ANALOG_BUTTON_AXIS], BUTTON_4_LOWERBAND,BUTTON_4_UPPERBAND);
    m_analogButtons[GAMEPAD_BUTTON_5] = new AnalogButton( m_axis[RIGHT_ANALOG_BUTTON_AXIS], BUTTON_5_LOWERBAND,BUTTON_5_UPPERBAND);
    m_analogButtons[GAMEPAD_BUTTON_6] = new AnalogButton(m_axis[LEFT_ANALOG_BUTTON_AXIS], BUTTON_6_LOWERBAND,BUTTON_6_UPPERBAND);
    m_analogButtons[GAMEPAD_BUTTON_7] = new AnalogButton( m_axis[RIGHT_ANALOG_BUTTON_AXIS], BUTTON_7_LOWERBAND,BUTTON_7_UPPERBAND);
    m_analogButtons[GAMEPAD_BUTTON_8] = new AnalogButton(m_axis[LEFT_ANALOG_BUTTON_AXIS], BUTTON_8_LOWERBAND,BUTTON_8_UPPERBAND);
    m_analogButtons[GAMEPAD_BUTTON_9] = new AnalogButton( m_axis[RIGHT_ANALOG_BUTTON_AXIS], BUTTON_9_LOWERBAND,BUTTON_9_UPPERBAND);
    m_analogButtons[GAMEPAD_BUTTON_10] = new AnalogButton(m_axis[LEFT_ANALOG_BUTTON_AXIS], BUTTON_10_LOWERBAND,BUTTON_10_UPPERBAND);
    m_analogButtons[GAMEPAD_BUTTON_11] = new AnalogButton( m_axis[RIGHT_ANALOG_BUTTON_AXIS], BUTTON_11_LOWERBAND,BUTTON_11_UPPERBAND);
    m_analogButtons[GAMEPAD_BUTTON_12] = new AnalogButton(m_axis[LEFT_ANALOG_BUTTON_AXIS], BUTTON_12_LOWERBAND,BUTTON_12_UPPERBAND);
    m_analogButtons[GAMEPAD_BUTTON_13] = new AnalogButton( m_axis[RIGHT_ANALOG_BUTTON_AXIS], BUTTON_13_LOWERBAND,BUTTON_13_UPPERBAND);
    //m_analogButtons[GAMEPAD_BIG_RED_BUTTON] = new AnalogButton(m_gamepad, GAMEPAD_BIG_RED_BUTTON,);

    m_button[GAMEPAD_SWITCH_18] = new DigitalButton(m_gamepad, 1);
    m_button[GAMEPAD_SWITCH_19] = new DigitalButton(m_gamepad, 2);
    m_button[GAMEPAD_SWITCH_20] = new DigitalButton(m_gamepad, 3);
    m_button[GAMEPAD_SWITCH_21] = new DigitalButton(m_gamepad, 4);
    m_button[GAMEPAD_BUTTON_14_UP] = new DigitalButton(m_gamepad, 7);
    m_button[GAMEPAD_BUTTON_14_DOWN] = new DigitalButton(m_gamepad, 6);
    m_button[GAMEPAD_BUTTON_15_UP] = new DigitalButton(m_gamepad, 9);
    m_button[GAMEPAD_BUTTON_15_DOWN] = new DigitalButton(m_gamepad, 8);
}

DragonGamepad::~DragonGamepad()
{
    delete m_gamepad;
    m_gamepad = nullptr;
}

double DragonGamepad::GetAxisValue
(
    AXIS_IDENTIFIER axis
) const
{
    double output = 0.0;
	if ( axis < MAX_GAMEPAD_AXIS )
	{
		switch ( axis )//switch method to find out the axes for each gamepad buttons that use axis.
		{
		case GAMEPAD_AXIS_16:
			output = m_gamepad->GetY( frc::GenericHID::kLeftHand ) * m_axisInversionFactor[ axis ];
			break;

		case GAMEPAD_AXIS_17:
			output = m_gamepad->GetY( frc::GenericHID::kRightHand ) * m_axisInversionFactor[ axis ];
			break;

		// jw - changed 2 to axis
		case GAMEPAD_DIAL_22: //Dial has six settings, 22-27.
			output = m_gamepad->GetY( frc::GenericHID::kLeftHand ) * m_axisInversionFactor[ axis ];
			break;

		case GAMEPAD_DIAL_23:
			output = m_gamepad->GetY( frc::GenericHID::kLeftHand ) * m_axisInversionFactor[ axis ];
			break;

		case GAMEPAD_DIAL_24:
			output = m_gamepad->GetY( frc::GenericHID::kLeftHand ) * m_axisInversionFactor[ axis ];
			break;

		case GAMEPAD_DIAL_25:
			output = m_gamepad->GetY( frc::GenericHID::kLeftHand ) * m_axisInversionFactor[ axis ];
			break;

		case GAMEPAD_DIAL_26:
			output = m_gamepad->GetY( frc::GenericHID::kLeftHand ) * m_axisInversionFactor[ axis ];
			break;

		case GAMEPAD_DIAL_27:
			output = m_gamepad->GetY( frc::GenericHID::kLeftHand ) * m_axisInversionFactor[ axis ];
			break;

		default:
			break;

		}
		if (std::abs(output) < APPLY_STANDARD_DEADBAND)
		{
			output = 0.0;
		}
		else if (std::abs(output)> APPLY_STANDARD_DEADBAND)
		{
			if ( m_axisProfile[axis] == CUBED )
			{
				output = pow( output, 3.0 );
			}
			output = output * m_axisScale[axis];
		}
	}
	return output;
}

bool DragonGamepad::IsButtonPressed
(
    BUTTON_IDENTIFIER button
) const
{
    bool isPressed = false;
    return isPressed;
}

void DragonGamepad::SetAxisDeadband
(
    AXIS_IDENTIFIER axis,
    AXIS_DEADBAND type
)
{
    if ( m_axis[axis] != nullptr )
    {
        m_axis[axis]->SetDeadBand( type );
    }        
}

void DragonGamepad::SetAxisProfile
(
    AXIS_IDENTIFIER axis,
    AXIS_PROFILE curve
)
{
    if ( m_axis[axis] != nullptr )
    {
        m_axis[axis]->SetAxisProfile( curve );
    }        
}

void DragonGamepad::SetAxisScale
(
    AXIS_IDENTIFIER axis,
    float scaleFactor
)
{
    if ( m_axis[axis] != nullptr )
    {
        m_axis[axis]->SetAxisScaleFactor( scaleFactor );
    }        
}

void DragonGamepad::SetButtonMode
(
    BUTTON_IDENTIFIER button,
    BUTTON_MODE mode
) 
{

}

double DragonGamepad::GetRawAxis
(
    AXIS_IDENTIFIER axis
)
{
    double value = 0.0;
    if ( m_axis[axis] != nullptr )
    {
        value = m_axis[axis]->GetAxisValue();
    }
    else
    {
        printf( "==>> no axis %d \n", axis );
    }
    return value;
}

bool DragonGamepad::IsRawButtonPressed
(
    BUTTON_IDENTIFIER button
) 
{
    bool pressed = false;
    if(m_button[button] != nullptr)
    {
        pressed = m_button[button]->IsButtonPressed();
    }
    return pressed;
}

bool DragonGamepad::IsRawButtonPressed
(
    ANALOG_BUTTON_IDENTIFIER button
) 
{
    bool pressed = false;
    if(m_analogButtons[button] != nullptr)
    {
        pressed = m_analogButtons[button]->IsButtonPressed();
    }
    return pressed;
}
int DragonGamepad::GetPOVValue
(
)
{
    return 0;
}

bool DragonGamepad::WasButtonPressed
(
    BUTTON_IDENTIFIER button
) const
{
    return false;
}

bool DragonGamepad::WasButtonReleased
(
    BUTTON_IDENTIFIER button
) const
{
    return false;
}